### PR TITLE
feat(swarm): deprecate `NegotiatedSubstream` in favor of `Stream`

### LINKED ISSUE #3759: core: encapsulate `multistream-select`
## Description

At the moment, the multiple types from `multistream-select` leak into the `libp2p-core` APIs, resulting in cascading breaking changes.

- Error types should be mapped to a new error in `libp2p-core`. This would also flatten the `UpgradeError` type.
- The resulting stream can be wrapped in a newtype.

## Motivation

Stop the cascade of breaking changes. For example, https://github.com/libp2p/rust-libp2p/issues/2831 could have been done in a backwards-compatible way if we would already have this in place.

## Are you planning to do it yourself in a pull request?

Yes.

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

This patch tackles two things at once that are fairly intertwined:

1. There is no such thing as a "substream" in libp2p, the spec and other implementations only talk about "streams". We fix this by deprecating `NegotiatedSubstream`.
2. Previously, `NegotiatedSubstream` was a type alias that pointed to a type from `multistream-select`, effectively leaking the version of `multistream-select` to all dependencies of `libp2p-swarm`. We fix this by introducing a `Stream` newtype.

Resolves: #3759.
Related: #3748.

## Notes & open questions

Despite being an invasive and technically breaking change, I think the upgrade should be pretty painless.

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: swarm/src/connection.rs [hunk 1]
```diff
@@ -470,24 +472,23 @@ impl<'a> IncomingInfo<'a> {
     }
 }
 
-struct SubstreamUpgrade<UserData, Upgrade> {
+struct SubstreamUpgrade<UserData, TOk, TErr> {
```
Comment (mxinden) [inline swarm/src/connection.rs]:
```suggestion
struct StreamUpgrade<UserData, TOk, TErr> {
```

Given that this is an internal type, why not already rename it here?

File: swarm/CHANGELOG.md [hunk 2]
```diff
@@ -42,6 +42,9 @@
   
   See [PR 3651].
 
+- Deprecate the `NegotiatedSubstream` type and replace it with `Stream`.
+  See [PR XXXX].
```
Comment (mxinden) [inline swarm/CHANGELOG.md]:
```suggestion
  See [PR 3912].
```

File: swarm/CHANGELOG.md [hunk 3]
```diff
@@ -51,6 +54,7 @@
 [PR 3884]: https://github.com/libp2p/rust-libp2p/pull/3884
 [PR 3885]: https://github.com/libp2p/rust-libp2p/pull/3885
 [PR 3886]: https://github.com/libp2p/rust-libp2p/pull/3886
+[PR XXXX]: https://github.com/libp2p/rust-libp2p/pull/XXXX
```
Comment (mxinden) [inline swarm/CHANGELOG.md]:
```suggestion
[PR 3912]: https://github.com/libp2p/rust-libp2p/pull/3912
```

**Review by mxinden — APPROVED** [review]
One small nit pick. Otherwise looks good to me. Good to get rid of `Substream*`!

File: swarm/src/connection.rs (hunk 1 above)
Comment (thomaseizinger) [reply swarm/src/connection.rs]: It didn't even occur to me, yes makes a lot of sense :)

Comment (thomaseizinger) [reply swarm/src/connection.rs]: There are a fair few more usages of "substream" but I think those can easily go in separate PRs.

Comment (mxinden) [reply swarm/src/connection.rs]: > There are a fair few more usages of "substream" but I think those can easily go in separate PRs.

:+1: works for me.

**Comment by drHuangMHT** [comment]
What about `SubstreamProtocol`? That's what you need to negotiate for an outbound (sub)stream, am I right?

**Comment by thomaseizinger** [comment]
> What about `SubstreamProtocol`? That's what you need to negotiate for an outbound (sub)stream, am I right?

We should get rid of all the "sub"-stream terminology yes. I've not had time to follow up on this PR. Most of the symbols are in individual crates so it is not urgent.

Perhaps the only one we should rename as part of this breaking release is `Muxer::Substream`.

**Pull request merged**